### DISCUSSION TITLE
How to speed up HinSAGE Model training for binary link prediction?

### DISCUSSION THREAD

**Discussion by kwanchet** [stellargraph/stellargraph#1795] [General]
Hi,

I have tried to do the following codes to split the data into train and test data. To give you a sense of my data, please see the following Stellargraph structure. After splitting the data, it took 12 hours for the HinSAGE model to complete an epoch of training within Azure notebook (see below for Azure configuration). It seems to be taking a long time. So my question is (i) is there any issue with the way I split the data or (ii) is there another way I can speed up the training?

# Stellargraph structure:
```
StellarGraph: Undirected multigraph
 Nodes: 127566, Edges: 156522

 Node types:
  eater: [127379]
    Features: float32 vector, length 31
    Edge types: eater-link->merchant
  merchant: [187]
    Features: float32 vector, length 146
    Edge types: merchant-link->eater

 Edge types:
    eater-link->merchant: [156522]
        Weights: all 1 (default)
        Features: none
```

# Azure Configuration:
![image](https://user-images.githubusercontent.com/24635154/91158713-a48a6580-e6f9-11ea-859a-fc66f1955438.png)

# train-test split code:
```
train_test_ratio = 0.05

# create reduce graph and edges for test data
edges_rg_test, edges_test = model_selection.train_test_split(
    pos_edge_df, train_size= 1 - train_test_ratio , test_size=train_test_ratio
)

# create reduce graph and edges for train data
edges_rg_train, edges_train = model_selection.train_test_split(
    edges_rg_test, train_size= 1 - train_test_ratio , test_size=train_test_ratio
)
  
# convert test and train reduce graphs into stellargraph format
graph_test = StellarGraph({"eater": eater_nodes, "merchant": merchant_nodes}, edges_rg_test)
graph_train = StellarGraph({"eater": eater_nodes, "merchant": merchant_nodes}, edges_rg_train)

# convert ids into series
eater_ids = pos_edge_df.source.to_numpy()
merchant_ids = pos_edge_df.target.to_numpy()

# generate all possible edges
all_possible_edges = pd.DataFrame(
    {
        "source": np.random.choice(eater_ids, size=200000),
        "target": np.random.choice(merchant_ids, size=200000),
        "weight": 0
    }
)

# generate negative edges 
all_possible_edges = pd.concat([all_possible_edges,pos_edge_df])
all_possible_edges = all_possible_edges.groupby(['source', 'target'])['weight'].agg('sum').reset_index()
neg_edge_df = all_possible_edges.loc[all_possible_edges['weight'] == 0]


# sample the negative edge cases for test (same number as positive edge cases)
neg_test_ratio = len(edges_test)/len(neg_edge_df)

edges_neg, edges_neg_test = model_selection.train_test_split(
    neg_edge_df, train_size=(1.0 - neg_test_ratio), test_size= neg_test_ratio
)

# sample the negative edge cases for train (same number as positive edge cases)
neg_train_ratio = len(edges_train)/len(edges_neg)

edges_neg, edges_neg_train = model_selection.train_test_split(
    edges_neg, train_size=(1.0 - neg_train_ratio), test_size= neg_train_ratio
)

# combine pos and neg cases to train and test
edges_test = pd.concat([edges_test,edges_neg_test])
edges_train = pd.concat([edges_train,edges_neg_train])

# create edgelist and labels
examples_train = list(edges_train[["source", "target"]].itertuples(index=False))
examples_test = list(edges_test[["source", "target"]].itertuples(index=False))

labels_train = edges_train["weight"]
labels_test = edges_test["weight"]
```

**Comment by huonw** [comment]
Hi, thanks for asking a question.

How big is your graph (e.g. `print(graph.info())` for each of them? How do you construct the generators (e.g. what's the batch size)?

**Comment by kwanchet** [comment]
@huonw thanks for the reply. great to hear from you again. hope things are good for you!

I have managed to find the problem as I have used the batch size of 20,000. It works fine now after changing it to 500. 

On a separate note, we have trained the model using a different combinations of hyper-parameters but the AUC does not increase beyond 0.82. Please see below for our best model hyper-parameters:

<img width="1142" alt="Screenshot 2020-09-21 at 9 03 25 PM" src="https://user-images.githubusercontent.com/24635154/93769883-0b923000-fc4e-11ea-85fe-bce6874d097a.png">

Is there a way that I can increase the AUC?

**Reply by huonw** [reply]
That's great. With HinSAGE especially, a large batch size like that likely involves creating and copying huge arrays and tensors, which may exhaust memory.

As for increasing AUC, that likely depends a lot on your specific problem. It can be influenced by things like:

- any data preprocessing and feature preparation you perform
- the number of layers in the model
- the size of each layer
- the type of layers and activations (especially for the post-HinSAGE classification layers)

Without knowing more specifics, I can't offer specific advice.

**Reply by kwanchet** [reply]
Hi @huonw ,

i have preprocessed both edge and node dataframes. I ensure that there are unique records within node dataframes together with their features (or attributes) while the edge dataframe mainly contains the ids of eater and merchant with the weight = 1 (to represent true positive edge). There are 2 hidden layers and their sizes are [144,72]. Both tanh and sigmoid functions are used as activation functions.  

I have tried tuning epochs (to 250), NUM_SAMPLES (to [8,8]), layer sizes to [144,72]. But it does not help to increase the AUC to 0.89. For info, previously, i ran the model on 0.0009 of the dataset (i.e. a very small dataset) and I was able to obtain AUC of 0.89. Currently, I am training the model on 0.20 of the same dataset. I suspect that it is harder to train the model given that the dataset size has increased and the sparsity of the dataset (i.e. number of edges (156522) is slightly greater than number of nodes (127566)).


below are the specifics requested:

```

BIAS = True
ACTIVATIONS = ["tanh", "sigmoid"]
EDGE_EMBEDDING_METHOD = "concat"
BATCH_SIZE = 2000
NUM_SAMPLES = [3,2] #number of neighbour at each layer
LEARNING_RATE = 0.001
DROPOUT = 0.3
LAYER_SIZES = [144,72] 
EPOCHS = 30

model.compile( optimizer=optimizers.Adam(lr=LEARNING_RATE), loss=keras.losses.binary_crossentropy, metrics=METRICS, )


Model: "functional_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input_3 (InputLayer)            [(None, 3, 146)]     0                                            
__________________________________________________________________________________________________
input_5 (InputLayer)            [(None, 6, 31)]      0                                            
__________________________________________________________________________________________________
input_6 (InputLayer)            [(None, 6, 146)]     0                                            
__________________________________________________________________________________________________
input_1 (InputLayer)            [(None, 1, 31)]      0                                            
__________________________________________________________________________________________________
reshape (Reshape)               (None, 1, 3, 146)    0           input_3[0][0]                    
__________________________________________________________________________________________________
reshape_2 (Reshape)             (None, 3, 2, 31)     0           input_5[0][0]                    
__________________________________________________________________________________________________
input_4 (InputLayer)            [(None, 3, 31)]      0                                            
__________________________________________________________________________________________________
reshape_3 (Reshape)             (None, 3, 2, 146)    0           input_6[0][0]                    
__________________________________________________________________________________________________
dropout_1 (Dropout)             (None, 1, 31)        0           input_1[0][0]                    
__________________________________________________________________________________________________
dropout (Dropout)               (None, 1, 3, 146)    0           reshape[0][0]                    
__________________________________________________________________________________________________
dropout_5 (Dropout)             (None, 3, 146)       0           input_3[0][0]                    
__________________________________________________________________________________________________
dropout_4 (Dropout)             (None, 3, 2, 31)     0           reshape_2[0][0]                  
__________________________________________________________________________________________________
input_2 (InputLayer)            [(None, 1, 146)]     0                                            
__________________________________________________________________________________________________
reshape_1 (Reshape)             (None, 1, 3, 31)     0           input_4[0][0]                    
__________________________________________________________________________________________________
dropout_7 (Dropout)             (None, 3, 31)        0           input_4[0][0]                    
__________________________________________________________________________________________________
dropout_6 (Dropout)             (None, 3, 2, 146)    0           reshape_3[0][0]                  
__________________________________________________________________________________________________
mean_hin_aggregator (MeanHinAgg multiple             12888       dropout_1[0][0]                  
                                                                 dropout[0][0]                    
                                                                 dropout_7[0][0]                  
                                                                 dropout_6[0][0]                  
__________________________________________________________________________________________________
mean_hin_aggregator_1 (MeanHinA multiple             12888       dropout_3[0][0]                  
                                                                 dropout_2[0][0]                  
                                                                 dropout_5[0][0]                  
                                                                 dropout_4[0][0]                  
__________________________________________________________________________________________________
dropout_3 (Dropout)             (None, 1, 146)       0           input_2[0][0]                    
__________________________________________________________________________________________________
dropout_2 (Dropout)             (None, 1, 3, 31)     0           reshape_1[0][0]                  
__________________________________________________________________________________________________
reshape_4 (Reshape)             (None, 1, 3, 144)    0           mean_hin_aggregator_1[1][0]      
__________________________________________________________________________________________________
reshape_5 (Reshape)             (None, 1, 3, 144)    0           mean_hin_aggregator[1][0]        
__________________________________________________________________________________________________
dropout_9 (Dropout)             (None, 1, 144)       0           mean_hin_aggregator[0][0]        
__________________________________________________________________________________________________
dropout_8 (Dropout)             (None, 1, 3, 144)    0           reshape_4[0][0]                  
__________________________________________________________________________________________________
dropout_11 (Dropout)            (None, 1, 144)       0           mean_hin_aggregator_1[0][0]      
__________________________________________________________________________________________________
dropout_10 (Dropout)            (None, 1, 3, 144)    0           reshape_5[0][0]                  
__________________________________________________________________________________________________
mean_hin_aggregator_2 (MeanHinA (None, 1, 72)        10440       dropout_9[0][0]                  
                                                                 dropout_8[0][0]                  
__________________________________________________________________________________________________
mean_hin_aggregator_3 (MeanHinA (None, 1, 72)        10440       dropout_11[0][0]                 
                                                                 dropout_10[0][0]                 
__________________________________________________________________________________________________
reshape_6 (Reshape)             (None, 72)           0           mean_hin_aggregator_2[0][0]      
__________________________________________________________________________________________________
reshape_7 (Reshape)             (None, 72)           0           mean_hin_aggregator_3[0][0]      
__________________________________________________________________________________________________
lambda (Lambda)                 (None, 72)           0           reshape_6[0][0]                  
                                                                 reshape_7[0][0]                  
__________________________________________________________________________________________________
link_embedding (LinkEmbedding)  (None, 144)          0           lambda[0][0]                     
                                                                 lambda[1][0]                     
__________________________________________________________________________________________________
dense (Dense)                   (None, 1)            145         link_embedding[0][0]             
__________________________________________________________________________________________________
reshape_8 (Reshape)             (None, 1)            0           dense[0][0]                      
==================================================================================================
Total params: 46,801
Trainable params: 46,801
Non-trainable params: 0

```
